[SourceKit] Allow explicit cancellation of requests with a cancellation token

ahoppen opened the pull request:
The key changes here are
- To keep track of cancellation tokens for all `ScheduledConsumer`s in `SwiftASTManager`
- Generate unique request handles for all incoming requests (`create_request_handle `), use these request handles as cancellation tokens and return them from the `sourcekitd_send_request` methods
- Implement cancellation with `sourcekitd_cancel_request` as the entry point and `SwiftASTManager::cancelASTConsumer` as the termination point

Everything else is just plumbing the cancellation token through the various abstraction layers.

rdar://83391505

ahoppen committed:
[SourceKit] Allow explicit cancellation of requests with a cancellation token

The key changes here are
- To keep track of cancellation tokens for all `ScheduledConsumer`s in `SwiftASTManager`
- Generate unique request handles for all incoming requests (`create_request_handle `), use these request handles as cancellation tokens and return them from the `sourcekitd_send_request` methods
- Implement cancellation with `sourcekitd_cancel_request` as the entry point and `SwiftASTManager::cancelASTConsumer` as the termination point

Everything else is just plumbing the cancellation token through the various abstraction layers.

rdar://83391505

ahoppen commented:
@swift-ci Please smoke test

benlangmuir commented on tools/SourceKit/tools/sourcekitd/bin/InProc/sourcekitdInProc.cpp:
```diff
@@ -111,14 +111,24 @@ void sourcekitd::set_interrupted_connection_handler(
 
+/// Create a new SourceKit request handle. Each call of this method is
+/// guaranteed to return a new, unique handle.
+sourcekitd_request_handle_t create_request_handle(void) {
```
Can this function be `static`?

ahoppen replied:
Could idea. I hadn’t thought about that. 👍

benlangmuir commented on tools/SourceKit/tools/sourcekitd/bin/InProc/sourcekitdInProc.cpp:
```diff
@@ -111,14 +111,24 @@ void sourcekitd::set_interrupted_connection_handler(
+/// guaranteed to return a new, unique handle.
+sourcekitd_request_handle_t create_request_handle(void) {
+  static std::atomic<size_t> handle(1);
+  return reinterpret_cast<sourcekitd_request_handle_t>(handle.fetch_add(1));
```
Nit: could be relaxed memory order since we only care about it being unique.

ahoppen replied:
Thanks. Good idea. Forgot about the relaxed memory order we use everywhere else.

benlangmuir commented on tools/SourceKit/tools/sourcekitd/bin/XPC/Client/sourcekitd.cpp:
```diff
@@ -106,6 +106,13 @@ sourcekitd_set_notification_handler(sourcekitd_response_receiver_t receiver) {
 
+/// Create a new SourceKit request handle. Each call of this method is
+/// guaranteed to return a new, unique handle.
+sourcekitd_request_handle_t create_request_handle(void) {
```
Same comments RE: `static` and `fetch_add`.

benlangmuir commented on tools/SourceKit/tools/sourcekitd/bin/XPC/Client/sourcekitd.cpp:
```diff
@@ -172,6 +183,9 @@ void sourcekitd_send_request(sourcekitd_object_t req,
   xpc_object_t msg = xpc_dictionary_create(nullptr, nullptr,  0);
   xpc_dictionary_set_value(msg, "msg", contents);
+  if (request_handle) {
+    xpc_dictionary_set_value(msg, "cancelToken", request_handle);
```
* `xpc_dictionary_set_value` expects an `xpc_object_t`. I think you want `xpc_dictionary_set_uint64`
* This string constant could go in `Internal-XPC.h` next to `KeyInternalMsg`.
* Nit: we've used underscores instead of camelCase for these names.

ahoppen replied:
> * `xpc_dictionary_set_value` expects an `xpc_object_t`. I think you want `xpc_dictionary_set_uint64`

I thought because `xpc_object_t`’s definition is `typedef void * xpc_object_t`, `xpc_object_t` would just be an opaque pointer (which is what we are using for the cancellation token as well). Or am I making a wrong assumption here?

> * This string constant could go in `Internal-XPC.h` next to `KeyInternalMsg`.

I guess the `"msg"` string should also go into `Internal-XPC.h` as well in that case, right?

> * Nit: we've used underscores instead of camelCase for these names.

👍

benlangmuir replied:
> I thought because xpc_object_t’s definition is typedef void * xpc_object_t, xpc_object_t would just be an opaque pointer (which is what we are using for the cancellation token as well). Or am I making a wrong assumption here?

`xpc_object_t` is `void *` because it's an opaque type in C, but it needs to be a real object created using one of the APIs like `xpc_dictionary_create()`.  Passing in an arbitrary pointer will crash.

We should probably add at least one test for this that uses `sourcekitd-test` so that we can catch this kind of issue.

ahoppen replied:
Ah, that makes sense. I also added cancellation support to `sourcekitd-test` now so we should be exercising that code path if `sourcekitd-test` is built with XPC.

benlangmuir commented on tools/SourceKit/tools/sourcekitd/bin/XPC/Client/sourcekitd.cpp:
```diff
@@ -201,7 +215,13 @@ void sourcekitd_send_request(sourcekitd_object_t req,
-  // FIXME: Implement cancelling.
+  xpc_connection_t conn = getGlobalConnection();
+  xpc_object_t msg = xpc_dictionary_create(nullptr, nullptr, 0);
+  xpc_dictionary_set_value(msg, "cancel", handle);
```
Same issue with `xpc_dictionary_set_value` vs. uint64.

benlangmuir commented on tools/SourceKit/tools/sourcekitd/bin/XPC/Service/XPCService.cpp:
```diff
@@ -248,28 +248,32 @@ static void sourcekitdServer_peer_event_handler(xpc_connection_t peer,
-        contents = xpc_dictionary_get_value(event, "ping");
-        assert(contents && "unexpected message");
         xpc_object_t reply = xpc_dictionary_create_reply(event);
+        xpc_release(event);
```
Heh, nice catch.

benlangmuir commented on tools/SourceKit/tools/sourcekitd/bin/XPC/Service/XPCService.cpp:
```diff
@@ -248,28 +248,32 @@ static void sourcekitdServer_peer_event_handler(xpc_connection_t peer,
-      if (!contents) {
+      if (xpc_object_t contents = xpc_dictionary_get_value(event, "msg")) {
+        const void *cancelToken =
+            xpc_dictionary_get_value(event, "cancelToken");
```
xpc_dictionary_get_uint64

benlangmuir commented on unittests/SourceKit/SwiftLang/CursorInfoTest.cpp:
```diff
@@ -455,3 +459,40 @@ TEST_F(CursorInfoTest, CursorInfoCancelsPreviousRequest) {
+
+  open(SlowDocName, SlowContents, llvm::makeArrayRef(Args));
+
+  SourceKitCancellationToken CancellationToken = &CancellationToken;
```
A stack address is not necessarily unique in case we add more cancellation tests this could be problematic. Creating a global counter in this test file should be fine.

ahoppen commented:
Thanks for your feedback @benlangmuir.

I addressed the comments and also moved the `"msg"` to `Internal-XPC.h`. I’m still using `xpc_dictionary_set_value ` to set the cancellation token because
> I thought because xpc_object_t’s definition is typedef void * xpc_object_t, xpc_object_t would just be an opaque pointer (which is what we are using for the cancellation token as well). Or am I making a wrong assumption here?

ahoppen committed:
[SourceKit] Add an option to cancel async requests

ahoppen commented:
@swift-ci Please smoke test

benlangmuir commented on test/SourceKit/Misc/cancellation.swift:
```diff
@@ -0,0 +1,13 @@
+// We need to wait a little bit after request scheduling and cancellation to make sure we are not cancelling the request before it got scheduled.
+
+// RUN: not %sourcekitd-test -req=cursor -id=slow -async -pos=10:3 %s -- %s == \
+// RUN: -shell -- sleep 1 == \
```
I thought this would need `// REQUIRES: shell`, but I guess the tests passed on all platforms :shrug:

benlangmuir reviewed (approved):
LGTM!
